app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt

```diff
@@ -0,0 +1,115 @@
+package com.rockthevote.grommet.ui.eventFlow
+
+import androidx.lifecycle.*
+import com.rockthevote.grommet.data.api.RockyService
+import com.rockthevote.grommet.data.db.dao.PartnerInfoDao
+import com.rockthevote.grommet.data.db.model.PartnerInfo
+import com.rockthevote.grommet.util.coroutines.DispatcherProvider
+import com.rockthevote.grommet.util.coroutines.DispatcherProviderImpl
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+/**
+ * Created by Mechanical Man on 5/16/20.
+ */
+class PartnerLoginViewModel(
+        dispatchers: DispatcherProvider = DispatcherProviderImpl(),
+        private val rockyService: RockyService,
+        private val partnerInfoDao: PartnerInfoDao
+) : ViewModel() {
+
+    val partnerInfoId: LiveData<Long> =
+            Transformations.map(partnerInfoDao.getCurrentPartnerInfo()) { result ->
+                result.partnerInfoId
+            }
+
+    private val _partnerLoginState = MutableLiveData<PartnerLoginState>(PartnerLoginState.Init)
+    val partnerLoginState: LiveData<PartnerLoginState> = _partnerLoginState
+
+    private val _effect = MutableLiveData<PartnerLoginState.Effect?>()
+    val effect: LiveData<PartnerLoginState.Effect?> = _effect
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+        // TODO Should we handle this another way? Is it needed?
+        Timber.e(throwable)
+    }
+
+    private val supervisorJob = SupervisorJob()
+
+    private val rockyRequestScope = CoroutineScope(dispatchers.io + coroutineExceptionHandler)
```

Comment:
This is one of the places I made a mistake in my PR. You need to add the supervisor job to this, then make sure you clear it in `onCleared`

Reply:
eg. `private val rockyRequestScope = CoroutineScope(dispatchers.io + coroutineExceptionHandler + supervisorJob)` then
```
    override fun onCleared() {
        super.onCleared()
        supervisorJob.cancelChildren()
    }
```

Reply:
I thought you said that i don't need the supervisor job because i'm not doing parallel calls?

Reply:
can I just call`rockRequestScope.cancel()` ?

Reply:
> I thought you said that i don't need the supervisor job because i'm not doing parallel calls?

That's correct, but you still need a job associated with this if you're not using the `viewModelScope` in order to clean this up. `viewModelScope` automatically cleans itself up when the ViewModel is destroyed so we don't have to worry about cleaning it up.

In this case, I'd just get rid of `rockyRequestScope` all together, and use `viewModelScope` with `dispatchers.io` like you are in `init`

Reply:
👍